src/main/scala/cromiam/webservice/CromIamApiService.scala

```diff
@@ -44,186 +56,122 @@ trait CromIamApiService extends Directives with SprayJsonSupport with DefaultJso
     Http().singleRequest(cromwellRequest)
   }
 
+  private[webservice] def authorizeThenForwardToCromwell(user: Option[String], workflowId: String, action: String, request: HttpRequest): Future[HttpResponse] = {
+    (for {
+      _ <- requestAuth(WorkflowAuthorizationRequest(user.getOrElse("anon"), workflowId, action))
+      resp <- forwardToCromwell(request)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "")
+      case SamActorDown => HttpResponse(status = InternalServerError, entity = "")
+    }
+  }
+
+  private[webservice] def authorizeReadThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "read", request)
+  private[webservice] def authorizeAbortThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "abort", request)
+
+  private[webservice] def authorizeSubmitThenForwardToCromwell(user: Option[String], request: HttpRequest, batch: Boolean): Future[HttpResponse] = {
+
+    def extractWorkflowIds(response: HttpResponse): Future[List[String]] = {
+      if (batch) {
+        Unmarshal(response).to[List[CromwellStatus]].map(_.map(_.id))
+      } else {
+        Unmarshal(response).to[CromwellStatus].map(s => List(s.id))
+      }
+    }
+
+    val userNonOptional = user.getOrElse("anon")
+    (for {
+      resp <- forwardToCromwell(request)
+      workflowIds <- extractWorkflowIds(resp)
+      _ <- registerCreation(userNonOptional, workflowIds)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "{}")
+    }
+  }
+
   val workflowRoutes: Route = queryRoute ~ queryPostRoute ~ workflowOutputsRoute ~ submitRoute ~ submitBatchRoute ~
     workflowLogsRoute ~ abortRoute ~ metadataRoute ~ timingRoute ~ statusRoute ~ backendRoute
 
   val engineRoutes: Route = statsRoute ~ versionRoute
 
   val allRoutes: Route = workflowRoutes ~ engineRoutes
 
-  def statusRoute: Route =
-    path("api" / "workflows" / Segment / Segment / "status") { (version, possibleWorkflowId) =>
-      get {
-        optionalHeaderValueByName(UserIdHeader) { userId =>
-          extractRequest { req =>
-            complete {
-              logger.info(s"Forwarding status request from user=$userId")
-              forwardToCromwell(req)
-            }
+  private def handleRequest(f: (Option[String], HttpRequest) => ToResponseMarshallable): Route = {
+    optionalHeaderValueByName(UserIdHeader) { userId =>
+      toStrictEntity(300.millis) {
```

Comment:
Actually I'm not sure this is something that anyone will ever realistically change... 

unless there's strong objections, I'll leave this hard-coded for now